server/bin/pbench-server-put-shim-001.py

```diff
@@ -0,0 +1,176 @@
+
+import os
+import sys
+import glob
+import tempfile
+
+from logging import Logger
+from pathlib import Path
+from collections import namedtuple
+
+from pbench.agent.results import CopyResultTb
+from pbench.common.exceptions import BadConfig
+from pbench.common.logger import get_pbench_logger
+from pbench.common.utils import md5sum
+from pbench.server import PbenchServerConfig
+from pbench.server.report import Report
+
+
+_NAME_ = "pbench-server-prep-shim-002"
+
+Results = namedtuple("Results", ["nstatus", "ntotal", "ntbs"])
+
+
+def get_receive_dir(config: PbenchServerConfig, logger: Logger) -> Path:
+
+    # we are explicitly handling version-002 data in this shim
+    receive_dir_prefix = config.get("pbench-server", "pbench-receive-dir-prefix")
+    if not receive_dir_prefix:
+        logger.error("Failed: config value pbench-receive-dir-prefix pbench-server")
+        return None
+
+    receive_dir = Path(f"{receive_dir_prefix}-002").resolve()
+    if not receive_dir.is_dir():
+        logger.error("Failed: {} does not exist, or is not a directory", receive_dir)
+        return None
+
+    return receive_dir
+
+
+def process_tb(
+    config: PbenchServerConfig, logger: Logger, receive_dir: Path
+) -> Results:
+    """process_tb - receives a tar ball and copies it to the remote server
+                    using the server's HTTP PUT method
+
+        Args -
+            config -- PbenchServer config object
+            logger -- logger object to use when emitting log entries during
+                      operation
+            receive_dir -- directory where tarballs are received from agent
+
+    """
+
+    # Check for results that are ready for processing: version 002 agents
+    # upload the MD5 file as xxx.md5.check and they rename it to xxx.md5
+    # after they are done with MD5 checking so that's what we look for.
+    list_check = glob.glob(
+        os.path.join(receive_dir, "**", "*.tar.xz.md5"), recursive=True
+    )
+
+    logger.info("{}", config.TS)
+    list_check.sort()
+    nstatus = ""
+
+    ntotal = ntbs = 0
+
+    # FIXME: keep a generated token which has longer duration
+
+    token = config.get("pbench-server", "put-token")
+
+    for tbmd5 in list_check:
+        ntotal += 1
+
+        # full pathname of tarball
+        tb = Path(tbmd5[0:-4])
+        tbmd5 = Path(tbmd5)
+        tbdir = tb.parent
+        controller = tbdir.name
+
+        try:
+            tarball_len, tarball_md5 = md5sum(tb)
+            crt = CopyResultTb(
+                controller, tb, tarball_len, tarball_md5, config, logger,
+            )
+            crt.copy_result_tb(token)
+        except Exception as e:
+            logger.error("Unexpected Error: {}", e)
+            continue
+
+        try:
+            os.remove(tbmd5)
+        except Exception as exc:
+            logger.error(
+                "{}: Warning: cleanup of successful copy operation failed: '{}'",
+                config.TS,
+                exc,
+            )
```

Comment:
What does the logger do with an `Exception` object?  That is, should we be explicitly stringifying the object (i.e., passing `str(exc)` at line 104)?

Reply:
Doesn't seem like we should be explicitly stringifying. logger does not seem to concatenate the messages. Is there any specific reason to propose this?

Reply:
You're specifying the exception object (`exc`) as an argument to the logger with a format placeholder which is to be filled in with information from the argument.  I just wanted to make sure that the resulting log message is clear and useful.  For instance, it would be disappointing if `exc` were formatted as `FileNotFoundError` instead of `No such file or directory: '/baddir/badfile.tar.xz.md5'`, especially if passing it as `str(exc)` would solve that problem.  (That is, I think we're counting on the exception formatting to report the name of the file that it is having trouble with, since we're not logging it explicitly ourselves, so I want to make sure that that happens.)

But, if it's good as is, I'm happy.

Reply:
I tried raising some exceptions and it doesn't seem to have caused any issue with that. I will still look once again if any such issue occur :)